cpp/src/transform/row_bit_count.cu

```diff
@@ -0,0 +1,508 @@
+    // if we're entering a new branch.
+    // NOTE: this case can happen (a pop and a push by the same column)
+    // when we have a struct<list, list>
+    if (info[idx].branch_depth_end > info[idx].branch_depth_start) {
+      branch[branch_depth++] = cur_span;
+    }
+
+    // if we're back at depth 0, this is a new top-level column, so reset
+    // span info
+    if (info[idx].depth == 0) {
+      branch_depth      = 0;
+      last_branch_depth = 0;
+      cur_span          = row_span{tid, tid + 1};
+    }
+
+    // add the contributing size of this row
+    size += cudf::type_dispatcher(col.type(), row_size_functor{}, col, cur_span);
+
+    // if this is a list column, update the working span from our offsets
+    if (col.type().id() == type_id::LIST) {
+      column_device_view const& offsets = col.child(lists_column_view::offsets_column_index);
+      auto const base_offset            = offsets.data<offset_type>()[col.offset()];
+      cur_span.row_start =
+        offsets.data<offset_type>()[cur_span.row_start + col.offset()] - base_offset;
+      cur_span.row_end = offsets.data<offset_type>()[cur_span.row_end + col.offset()] - base_offset;
+    }
+
+    last_branch_depth = info[idx].branch_depth_end;
+  }
+}
+
+}  // anonymous namespace
+
+/**
+ * @copydoc cudf::detail::row_bit_count
+ *
+ */
+std::unique_ptr<column> row_bit_count(table_view const& t,
+                                      rmm::cuda_stream_view stream,
+                                      rmm::mr::device_memory_resource* mr)
+{
+  // flatten the hierarchy and determine some information about it.
+  std::vector<cudf::column_view> cols;
+  std::vector<column_info> info;
+  hierarchy_info h_info;
+  flatten_hierarchy(t.begin(), t.end(), cols, info, h_info, stream);
+
+  // create output buffer and view
+  auto output = cudf::make_fixed_width_column(
+    data_type{type_id::INT32}, t.num_rows(), mask_state::UNALLOCATED, stream, mr);
+  mutable_column_view mcv = output->mutable_view();
+
+  // simple case.  if we have no complex types (lists, strings, etc), the per-row size is already
+  // trivially computed
+  if (h_info.complex_type_count <= 0) {
+    thrust::fill(rmm::exec_policy(stream),
+                 mcv.begin<size_type>(),
+                 mcv.end<size_type>(),
+                 h_info.simple_per_row_size);
+    return output;
+  }
+
+  // create a contiguous block of column_device_views
+  auto d_cols = contiguous_copy_column_device_views<column_device_view>(cols, stream);
+
+  // move stack info to the gpu
+  rmm::device_uvector<column_info> d_info(info.size(), stream);
+  CUDA_TRY(cudaMemcpyAsync(d_info.data(),
+                           info.data(),
+                           sizeof(column_info) * info.size(),
+                           cudaMemcpyHostToDevice,
+                           stream.value()));
+  CUDF_EXPECTS(info.size() == cols.size(), "Size/info mismatch");
```

Comment:
How does this happen?
Perhaps it should be checked right after `flatten_heirarchy()` above?
Is this like an assert or should it be mentioned in the doxygen header with a `@throw` tag.